tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/WriteSourceRepoProperties.cs

```diff
@@ -0,0 +1,260 @@
+                    {
+                        return ($"20{((datePart / 1000))}{((datePart % 1000) / 50):D2}{(datePart % 50):D2}.{buildPart}", releaseParts[0]);
+                    }
+                }
+            }
+            else
+            {
+                // finalized version number (x.y.z) - probably not our code
+                // VSTest, Application Insights, Newtonsoft.Json do this
+                return (DateTime.Now.ToString("yyyyMMdd.1"), string.Empty);
+            }
+
+            throw new FormatException($"Can't derive a build ID from version {version} (commit count {commitCount})");
+        }
+
+        private static string GetRepoNameOrDefault(Dependency dependency)
+        {
+            if (dependency.RepoName != null)
+            {
+                return dependency.RepoName;
+            }
+            var repoUrl = dependency.Uri;
+            if (repoUrl.EndsWith(".git"))
+            {
+                repoUrl = repoUrl.Substring(0, repoUrl.Length - ".git".Length);
+            }
+            return repoUrl.Substring(repoUrl.LastIndexOf("/") + 1);
+        }
+
+        private static string GetCommitCount(string gitDir, string hash)
+        {
+            return RunGitCommand($"rev-list --count {hash}", gitDir: gitDir);
+        }
+
+        private static string GetCommitDate(string gitDir, string hash)
+        {
+            return RunGitCommand($"log -1 --format=%cd --date=short", gitDir: gitDir);
+        }
+
+        private static string RunGitCommand(string command, string workTree = null, string gitDir = null)
+        {
+            // Windows Git requires these to be before the command
+            if (workTree != null)
+            {
+                command = $"--work-tree={workTree} {command}";
+            }
+            if (gitDir != null)
+            {
+                command = $"--git-dir={gitDir} {command}";
+            }
+            var psi = new ProcessStartInfo { FileName = "git", Arguments = command, RedirectStandardOutput = true, RedirectStandardError = true, UseShellExecute = false, CreateNoWindow = true };
+            var p = new Process { StartInfo = psi };
+            p.Start();
+            var output = p.StandardOutput.ReadToEnd();
+            var error = p.StandardError.ReadToEnd();
+            p.WaitForExit();
+            if (p.ExitCode != 0)
+            {
+                throw new InvalidOperationException($"git command {command} failed with exit code {p.ExitCode} and error {error ?? "<blank>"}");
+            }
+            return output.Trim();
+        }
+
+        private static string DeriveRepoGitDirPath(string gitDirPath, string repoUrl)
+        {
+            if (repoUrl.EndsWith(".git"))
+            {
+                repoUrl = repoUrl.Substring(0, repoUrl.Length - ".git".Length);
+            }
+
+            return Path.Combine(gitDirPath, $"{repoUrl.Substring(repoUrl.LastIndexOf("/") + 1)}.git");
```

Comment:
I don't think so; the GitDir path always depends on the repo URL and can't be overridden like the repo name can.

Reply:
It sounds like that's a reason you can't just call the method, but the innards look pretty similar. Specifically, I'm seeing `repoUrl.Length - ".git".Length` and `.Substring(repoUrl.LastIndexOf("/") + 1)` multiple times in this file. It's not like it's terrible as-is since it's pretty simple logic, just a little deja vu that throws me off.

Reply:
Good point, factored this out.